src/Evolution/Systems/GeneralizedHarmonic/Characteristics.cpp

```diff
@@ -0,0 +1,191 @@
+    const tnsr::i<DataVector, Dim, Frame>& normal) noexcept {
+  auto char_speeds =
+      make_with_value<typename Tags::CharacteristicSpeeds<Dim, Frame>::type>(
+          get(lapse), 0.);
+  const auto shift_dot_normal = get(dot_product(shift, normal));
+  get(get<::Tags::CharSpeed<Tags::UPsi<Dim, Frame>>>(char_speeds)) =
+      -(1. + get(gamma_1)) * shift_dot_normal;
+  get(get<::Tags::CharSpeed<Tags::UZero<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal;
+  get(get<::Tags::CharSpeed<Tags::UPlus<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal + get(lapse);
+  get(get<::Tags::CharSpeed<Tags::UMinus<Dim, Frame>>>(char_speeds)) =
+      -shift_dot_normal - get(lapse);
+
+  return char_speeds;
+}
+
+
+template <size_t Dim, typename Frame>
+typename Tags::EvolvedFieldsFromCharacteristicFields<Dim, Frame>::type
+EvolvedFieldsFromCharacteristicFieldsCompute<Dim, Frame>::function(
+    const Scalar<DataVector>& gamma_2,
+    const tnsr::aa<DataVector, Dim, Frame>& u_psi,
+    const tnsr::iaa<DataVector, Dim, Frame>& u_zero,
+    const tnsr::aa<DataVector, Dim, Frame>& u_plus,
+    const tnsr::aa<DataVector, Dim, Frame>& u_minus,
+    const tnsr::i<DataVector, Dim, Frame>& unit_normal_one_form) noexcept {
+  auto evolved_fields =
+      make_with_value<typename Tags::EvolvedFieldsFromCharacteristicFields<Dim,
+          Frame>::type>(get(gamma_2), 0.);
+
+  // Invert Eq.(32) of Lindblom+ (2005) for Psi
+  get<Tags::Psi<Dim, Frame>>(evolved_fields) = u_psi;
+
+  for (size_t mu = 0; mu < Dim + 1; ++mu) {
+    for (size_t nu = 0; nu < mu + 1; ++nu) {
+      // Invert Eq.(32) - (34) of Lindblom+ (2005) for Pi and Phi
+      get<Tags::Pi<Dim, Frame>>(evolved_fields).get(mu, nu) =
+                            0.5 * (u_plus.get(mu, nu) + u_minus.get(mu, nu)) +
+                            get(gamma_2) * u_psi.get(mu, nu);
+      for (size_t i = 0; i < Dim; ++i) {
+        get<Tags::Phi<Dim, Frame>>(evolved_fields).get(i,mu,nu) =
+                            0.5 * (u_plus.get(mu, nu) - u_minus.get(mu, nu)) *
+                            unit_normal_one_form.get(i) + u_zero.get(i, mu, nu);
+      }
+    }
+  }
+
+  return evolved_fields;
+}
+
+
+namespace {
+template <size_t Dim, typename Frame, typename DataType>
+struct CharacteristicFieldsComputeBuffer;
+
+template <size_t Dim, typename Frame>
+struct CharacteristicFieldsComputeBuffer<Dim, Frame, DataVector> {
+  private:
+    // We make one giant allocation so that we don't thrash the heap.
+    Variables<tmpl::list<::Tags::TempI<0, Dim, Frame>,
+                         ::Tags::Tempaa<1, Dim, Frame>,
+                         ::Tags::Tempiaa<2, Dim, Frame>>> buffer_;
+
+  public:
+    explicit CharacteristicFieldsComputeBuffer(const size_t size) noexcept :
+      buffer_(size, 0.), normal(get<::Tags::TempI<0, Dim, Frame>>(buffer_)),
+      phi_dot_normal(get<::Tags::Tempaa<1, Dim, Frame>>(buffer_)),
+      phi_dot_projection_tensor(get<::Tags::Tempiaa<2, Dim, Frame>>(buffer_)) {}
```

Comment:
I wonder if the extra allocation of phi_dot_projection_tensor costs more than just computing this every time you get UZero? No change for now, but maybe profile this down the road.

Reply:
will do so